File: Java/283_Move_Zeroes.java
// Time:O(1)
// Space:O(n)
// Updated 11/Aug

class Solution {
public void moveZeroes(int[] nums) {
    if (nums == null || nums.length == 0) return;        

    int insertPos = 0;
    for (int num: nums) {
        if (num != 0) nums[insertPos++] = num;
    }        

    while (insertPos < nums.length) {
        nums[insertPos++] = 0;
    }
}
}


///=============================

class Solution {
    public void moveZeroes(int[] nums) {
        for(int i = 0, j = 0; i < nums.length && j < nums.length; j++){
            if(nums[j] != 0){
                int temp = nums[i];
                nums[i++] = nums[j];
                nums[j] = temp;
            }
        }
    }
}

File: Python/283_move-zeroes.py
class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        left  = 0
        right = 1
        
        while left < len(nums) and right < len(nums):
            if nums[left] == 0:
                if nums[right] != 0:
                    nums[left], nums[right] = nums[right],nums[left]
                else:
                    right +=1
            else:
                left +=1
                right+=1